libs/ballot-interpreter/src/adjudication_reasons.ts

```diff
@@ -31,8 +35,83 @@ function compareMarkStatusDescending(
   return rankMarkStatus(markStatusB) - rankMarkStatus(markStatusA);
 }
 
+function getExpectedVoteCount(contest: AnyContest): number {
+  switch (contest.type) {
+    case 'candidate':
+      return contest.seats;
+    case 'yesno': // yes or no
+      return 1;
+    // istanbul ignore next
+    default:
+      throwIllegalValue(contest, 'type');
+  }
+}
+
 /**
- * Enumerates all the reasons a series of contests might need adjudication.
+ * Enumerates all the reasons a series of contests might need adjudication in
+ * the context of a BMD.
+ */
+export function getAllPossibleAdjudicationReasonsForBmdVotes(
+  contests: Contests,
+  votes: VotesDict
+): AdjudicationReasonInfo[] {
+  const reasons: AdjudicationReasonInfo[] = [];
+  let isBlankBallot = true;
+
+  for (const contest of contests) {
+    const expectedSelectionCount = getExpectedVoteCount(contest);
+    const actualVotes = assertDefined(votes[contest.id]);
+
+    const actualVoteCount = actualVotes.length;
+
+    if (isBlankBallot && actualVoteCount > 0) {
+      isBlankBallot = false;
+    }
+
+    if (actualVoteCount < expectedSelectionCount) {
+      const optionIds: string[] = [];
+      const contestType = contest.type;
+      switch (contestType) {
+        case 'candidate':
+          for (const option of actualVotes as CandidateVote) {
+            optionIds.push(option.id);
+          }
+          break;
+        case 'yesno':
+          for (const option of actualVotes as YesNoVote) {
+            // Should never be executed because a yes/no contest should have at
+            // most 1 vote so an undervote would be 0 votes. But the type
+            // doesn't enforce this, so we should handle the case where the
+            // contest has > 1 vote.
```

Comment:
Got it, I think I follow. The comment's wording took me a second to parse, but maybe something like this would be easier to follow:

> Because you can't overvote on a BMD, and the only way to undervote on a yes-no contest is to leave it blank, we'll never have any option IDs to populate here

Reply:
Clarified in [37fc820](https://github.com/votingworks/vxsuite/pull/5487/commits/37fc82086abf81e3d8d36c0afb36934234058b36) and removed no-op code

Reply:
Ah nice, good call

Reply:
At first I left the no-op code in because the type technically allowed for it, but on second thought I didn't like leaving in code that we never expect to execute and that has undefined business logic.